core/src/main/java/org/elasticsearch/search/query/QueryPhase.java

```diff
@@ -192,13 +199,49 @@ static boolean execute(SearchContext searchContext, final IndexSearcher searcher
 
             boolean timeoutSet = scrollContext == null && searchContext.timeout() != null &&
                 searchContext.timeout().equals(SearchService.NO_TIMEOUT) == false;
+
+            final Runnable timeoutRunnable;
             if (timeoutSet) {
-                // TODO: change to use our own counter that uses the scheduler in ThreadPool
-                // throws TimeLimitingCollector.TimeExceededException when timeout has reached
-                collectors.add(createTimeoutCollectorContext(searchContext.timeEstimateCounter(), searchContext.timeout().millis()));
+                final Counter counter = searchContext.timeEstimateCounter();
+                final long startTime = counter.get();
+                final long timeout = searchContext.timeout().millis();
+                final long maxTime = startTime + timeout;
+                timeoutRunnable = () -> {
+                    final long time = counter.get();
+                    if (time > maxTime) {
+                        queryResult.searchTimedOut(true);
+                        throw new CollectionTerminatedException();
```

Comment:
I don't think we should use the CollectionTerminatedException for this purpose. We have a special handling for this exception in the collectors but that's for the leaf level only. 
When the timeout is detected we should be able to stop the search immediately but if we have to build every scorer first it might be costly. Using a different exception that we catch at the higher level when we call `searcher.search` feels simpler to me and you don't need two levels of cancellation ?